keccak256/src/circuit.rs

```diff
@@ -47,217 +48,480 @@ impl<F: FieldExt> KeccakFConfig<F> {
-                // assignment
-                self.rho_config.assign_region(region, offset, out_state)?;
-                out_state
-            };
-            // Outputs in base-9 which is what Pi requires.
-            offset += RhoConfig::OFFSET;
-
-            // pi
-            state = {
-                // Apply pi outside circuit
-                let out_state = KeccakFArith::pi(&state_to_biguint(state));
-                let out_state = state_bigint_to_pallas(out_state);
                 // assignment
-                self.pi_config
-                    .assign_state(region, offset, state, out_state)?
+                self.rho_config.assign_rotation_checks(layouter, state)?
             };
+            // Outputs in base-9 which is what Pi requires
 
-            offset += PiConfig::OFFSET;
+            // Apply Pi permutation
+            state = pi_gate_permutation(state);
 
             // xi
             state = {
                 // Apply xi outside circuit
-                let out_state = KeccakFArith::xi(&state_to_biguint(state));
-                let out_state = state_bigint_to_pallas(out_state);
+                let out_state = KeccakFArith::xi(&state_to_biguint(split_state_cells(state)));
+                let out_state = state_bigint_to_field(out_state);
                 // assignment
-                self.xi_config
-                    .assign_state(region, offset, state, out_state)?
+                self.xi_config.assign_state(layouter, state, out_state)?
             };
 
-            offset += XiConfig::OFFSET;
+            // Last round before Mixing does not run IotaB9 nor BaseConversion
+            if round_idx == PERMUTATION - 1 {
+                break;
+            }
 
             // iota_b9
             state = {
-                let out_state = KeccakFArith::iota_b9(
-                    &state_to_biguint(state),
-                    ROUND_CONSTANTS[round],
-                );
-                let out_state = state_bigint_to_pallas(out_state);
+                let out_state =
+                    KeccakFArith::iota_b9(&state_to_biguint(split_state_cells(state)), *round_val);
+                let out_state = state_bigint_to_field(out_state);
                 self.iota_b9_config
-                    .not_last_round(region, offset, state, out_state, round)?;
-                out_state
+                    .not_last_round(layouter, state, out_state, round_idx)?
             };
-            offset += IotaB9Config::OFFSET;
-            // The resulting state is in Base-13 now. Which is what Theta
-            // requires again at the start of the loop.
 
-            self.not_mixing_b9_to_13.not_last_round();
+            // The resulting state is in Base-9 now. We now convert it to
+            // base_13 which is what Theta requires again at the
+            // start of the loop.
+            state = {
+                // TODO: That could be a Fixed column.
+                // Witness 1 for the activation flag.
```

Comment:
No, it's not as I was planning to convert this to a Fixed column so that it does not need to be witnessed nor constrained.

WDYT? If makes sense I can make the changes now.